### DISCUSSION TITLE
Image & Video Conversions

### DISCUSSION THREAD

**Discussion by EnochPrime** [immich-app/immich#1334] [General]
# Current State
This is the current state of affairs, to the best of my knowledge, as of `v1.41.1_64-dev`.

## Image Downscaling
When an image is uploaded the asset is stored as-is with no modifications. Image downscaling has been formal requested twice.
1. #1202 
2. #1242

These were ultimatly rejected on the grounds that the original asset is lost.
> Then if you want to retrieve the original file to print, you cannot do it anymore

> I don't know how I feel about it - storage is really cheap these days, ... and with Immich being primarily a backup type tool I expect most people would want to keep their original files.

## Video Trancoding
When a video is uploaded the asset is stored as-is, ~~unless it is not in an mp4 container. Then the video is transcoded per the server ffmpeg configuration settings. This functionality is request to go farther in #203 to convert the file when the codec is also not as configured. I believe this is done in the name of compatibility for web players.~~ An additional video file is created per the server ffmpeg configuration settings if the uploaded asset is not an mp4 container.

# The Proposal
## Issue to Solve
~~In my opinion the stance that has been taken for images and videos are contradicting. In the case of images the original is desired to be maintain, whereas with videos, undesired containers (and potentially codecs) are converted and thus the original asset is lost.~~ Nothing to solve, I guess, since there original assets are maintained. The below could be split into multiple feature requests.

## Video Conversions
Ultimately I believe the stance taken with images, that the original asset is desired, should hold true for videos as well. However, transcoding should not be removed entirely, and I believe there are a few cases that should be covered. In all of these scenarios new files would be generated from the original asset and stored alongside, much like thumbnails. Storage space concerns could be alleviated with on-demand transcoding.
- Playback compatibility: videos that don't meet the criteria for playback should be trancoded to something widely used. This could be done at full resolution or potentially a "high-quality" default such as the current 1280 by x.
- Data savings: playing videos over mobile data can be costly so one or many lower resolutions could be generated for playback in apps when on mobile data.
- Previews: just like thumbnails square aspect lower resolution videos should be generated for the timeline previews.

## Image Conversions
I believe these is also a case to be made for image conversions as well, but ultimately still keeping the original asset. Alternatively some of these could be generated on demand.
- heic to jpeg: not everyone supports this new format yet
- raw to jpeg: again jpeg is more universal
- Data savings: similar to videos, having lower resolution images to save mobile data
- Content protection: publicly shared albums could display low resolution images and not allow download of original assets

**Comment by bo0tzz** [comment]
> with videos, undesired containers (and potentially codecs) are converted and thus the original asset is lost.

This is incorrect. The transcoded video is stored separately, similar to the thumbnail files for images.

**Reply by EnochPrime** [reply]
I'm glad I misunderstood this! Thanks.

**Comment by alextran1502** [comment]
Am I understand correctly yet the discussion is to "whether or not having an on-demand video and image conversion?"

**Reply by EnochPrime** [reply]
Originally, no. It was more intended to be a discussion on what conversions should take place and why. All the conversions I suggest could take place on-upload if desired.

A separate discussion about what actions would trigger conversions would be warranted if any of the suggestions were to be implemented. This could be on a case-by-case basis.

**Reply by AngelaDMerkel** [reply]
Downscaling seems like a second best option in cases where losslessly encoding JPEGs as JXL can offer pretty big savings, including on older images which are likely to be lower resolutions. In the case of RAW files, stripping the preview JPEG in the RAW container and rendering a lower resolution thumbnail would also offer a 20% savings in my experience. Some manufacturers produces RAW files with substantial preview images in addition to the bayer data.

**Comment by AngelaDMerkel** [comment]
Since https://github.com/immich-app/immich/issues/1202 was closed, I thought this might be the best place to comment on offering lossless transcoding for some image types. Immich already support JXL which offers lossless re-encoding for JPEG files and savings as impressive as 30% or so. Users with large libraries which include TIFFs, JPEGS, and PNGs could all have their images losslessly rencoded and when a user downloads any of those images they can be bit-perfectly re-encoded to JPEG if users systems do not offer support for JXL.
Downscaling seems like an option but pretty significant savings can be made without the compromises one might associated with re-encoding.
